### PR TITLE
PLA-219 Add husky and a commit-msg hook

### PR DESCRIPTION
**Pull request by quesabe**

### CODE REVIEW

File: .husky/commit-msg [hunk 1]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+
+head -1 $@ | grep -Eq "^\w{3}-\d+\s.+" || \
```
Comment (gvidon) [inline .husky/commit-msg:4]: @quesabe why is it fixed to 3? Should we do `{2,4}`? I saw an evidence of all of such cases.

File: src/common/git/__tests__/index.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
+import * as fs from 'fs'
+import * as path from 'path'
+import {NodeProject, NodeProjectOptions} from 'projen/lib/javascript'
+import {synthSnapshot} from 'projen/lib/util/synth'
+import {addHusky, extendGitignore} from '..'
+
+describe('extendGitignore function', () => {
```
Comment (gvidon) [inline src/common/git/__tests__/index.ts:7]: @quesabe is this relevant to the PR?

File: src/common/git/assets/commit-msg [hunk 3]
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/env sh
+. "$(dirname -- "$0")/_/husky.sh"
+
+head -1 $@ | grep -Eq "^\w{3}-\d+\s.+" || \
```
Comment (gvidon) [inline src/common/git/assets/commit-msg:4]: @quesabe same as above.

File: src/common/git/husky.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
+/* eslint-disable import/no-relative-parent-imports -- JSII project rewrites tsconfig thus always overriding introduced aliases */
+import * as path from 'path'
+import type {NodeProject} from 'projen/lib/javascript'
+import {AssetFile} from '../files/AssetFile'
+import {WithGitHooks} from './with-git-hooks'
+
+export const addHusky = (project: NodeProject, options: WithGitHooks): void => {
+  if (options.hasGitHooks === false) {
```
Comment (gvidon) [inline src/common/git/husky.ts:8]: @quesabe can we make `false` a default?

File: src/common/git/husky.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    return
+  }
+
+  // ANCHOR Dependencies
```
Comment (gvidon) [inline src/common/git/husky.ts:12]: @quesabe I think adding these `ANCHOR`s don’t make any difference. Let’s remove them.

File: src/common/git/with-git-hooks.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+export interface WithGitHooks {
```
Comment (gvidon) [inline src/common/git/with-git-hooks.ts:1]: @quesabe let’s use `type` if that’s possible here.

File: src/apollo-server/index.ts [hunk 7]
```diff
@@ -122,6 +123,9 @@ export class OttofellerApolloServerProject extends TypeScriptAppProject {
       },
     })
 
+    // ANCHOR Setup git hooks with Husky
+    addHusky(this, options)
```
Comment (gvidon) [inline src/apollo-server/index.ts:127]: @quesabe I don’t really like the idea of checking the `hasHusky` option in `addJusky` function. Isn’t it parent caller’s responsibility? What do you think?

Comment (gvidon) [inline src/apollo-server/index.ts:127]: @quesabe I don’t really like the idea of checking the `hasHusky` option in `addJusky` function. Isn’t it parent caller’s responsibility? What do you think?

**Review by gvidon — CHANGES_REQUESTED** [review]

File: .husky/commit-msg (hunk 1 above)
Comment (quesabe) [reply .husky/commit-msg:4]: No problem. I just did not realise it might be something other than three letters.

File: src/common/git/__tests__/index.ts (hunk 2 above)
Comment (quesabe) [reply src/common/git/__tests__/index.ts:7]: Actually the extendGitignore tests are not added here. I moved the tests for extendGitignore and added the husky tests to the same file (since they are now under the `git` folder). Locally in _git_ it was marked as a move, but somehow GitHub does not recognise it and shows a file deletion followed by a file creation.

File: src/apollo-server/index.ts (hunk 7 above)
Comment (quesabe) [reply src/apollo-server/index.ts:127]: IMHO it's a matter of preference. I did it similar to other function. However it was done this way because of the `complexity` rule which did not allow too many conditionals within the constructor and I moved most of them inside the functions. For now the rule is gone and we are free to do it either way. For me it does not make much difference, so I'm ready to follow your lead :).

File: src/common/git/with-git-hooks.ts (hunk 6 above)
Comment (quesabe) [reply src/common/git/with-git-hooks.ts:1]: No, JSII noes not let it and fails on checking other interfaces that extend this one. And in case of replacing all the interfaces with type alias, we would eventually face the following error: `JSII1001: Non-primitive types without a symbol cannot be processed.`

File: src/common/git/__tests__/index.ts (hunk 2 above)
Comment (quesabe) [reply src/common/git/__tests__/index.ts:7]: Interestingly for me GitHub does not show all this lines as new - it shows the file as modified with the `addHusky` tests added.

File: src/apollo-server/index.ts (hunk 7 above)
Comment (quesabe) [reply src/apollo-server/index.ts]: Probably I need to review all the existing functions to move all such conditions outside them.

File: src/common/git/with-git-hooks.ts (hunk 6 above)
Comment (gvidon) [reply src/common/git/with-git-hooks.ts:1]: Got it, thanks.

File: .husky/commit-msg (hunk 1 above)
Comment (gvidon) [reply .husky/commit-msg]: Ok, thanks.

**Review by gvidon — APPROVED** [review]
